### ISSUE TITLE
Failed to compile on Next.js

### ISSUE THREAD

**Issue by kyleawayan** [pmndrs/react-postprocessing#72]
When I try to use this with Next.js,
```tsx
import {
  EffectComposer,
  DepthOfField,
  Bloom,
  Noise,
  Vignette,
} from "@react-three/postprocessing";

export default function ThreeDLogo() {
  return (
    <div style={{ height: "100vh" }}>
      <Canvas>
        <OrbitControls />
        <ambientLight intensity={0.6} />
        <Suspense fallback={null}>
          <Model />
        </Suspense>
        <EffectComposer>
          <DepthOfField
            focusDistance={0}
            focalLength={0.02}
            bokehScale={2}
            height={480}
          />
          <Bloom luminanceThreshold={0} luminanceSmoothing={0.9} height={300} />
          <Noise opacity={0.02} />
          <Vignette eskil={false} offset={0.1} darkness={1.1} />
        </EffectComposer>
      </Canvas>
    </div>
  );
}
```

with `ThreeDLogo.tsx` being imported like this:
```tsx
const ThreeDLogo = dynamic(() => import("../components/ThreeDLogo"), {
  ssr: false,
});

export default function Index() {
  return (
    <div>
      <ThreeDLogo />
    </div>
  );
}
```

I get this error:
```
Module parse failed: Unexpected token (1:2320)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> import{BlendFunction as e,BloomEffect as t,BrightnessContrastEffect as n,ChromaticAberrationEffect as r,ColorAverageEffect as i,ColorDepthEffect as o,DepthEffect as c,EffectComposer as s,RenderPass as l,NormalPass as a,EffectPass as u,DepthOfFieldEffect as d,DotScreenEffect as m,GlitchEffect as f,GlitchMode as p,GodRaysEffect as h,GridEffect as g,HueSaturationEffect as b,NoiseEffect as w,OutlineEffect as E,PixelationEffect as y,ScanlineEffect as v,SelectiveBloomEffect as S,SepiaEffect as x,SSAOEffect as T,SMAAPreset as j,EdgeDetectionMode as C,SMAAImageLoader as L,SMAAEffect as A,TextureEffect as R,ToneMappingEffect as P,VignetteEffect as z}from"postprocessing";import F,{forwardRef as O,useMemo as B,useLayoutEffect as D,createContext as M,useEffect as k,useRef as I,useImperativeHandle as N,useContext as G}from"react";import{Vector2 as _,HalfFloatType as H,Vector3 as Y,TextureLoader as q,sRGBEncoding as U...
```

Do I have to set any special settings for Webpack or anything?

**Comment by kyleawayan** [comment]
Using code from `Effects.js` from https://codesandbox.io/embed/r3f-game-i2160 for now

**Comment by nrbabcock** [comment]
I'm getting the same error

**Comment by mfanuzzi** [comment]
Fails on Gatsby as well. Presumably it's an issue with SSG; eg using `window` without checking for its existence.

**Comment by sargunv** [comment]
I'm using CRA, not Next, but I had a similar error and was able to resolve it by adding this Babel plugin: `@babel/plugin-proposal-optional-chaining`. A quick look at the index.js file in [the tarball](https://registry.npmjs.org/@react-three/postprocessing/-/postprocessing-1.5.2.tgz) shows optional chaining is being used in a couple spots.

**Comment by vinkovsky** [comment]
Same here

**Comment by MichalPrzybys-TomTom** [comment]
As suggested here https://github.com/webpack/webpack/issues/10227, there is a workaround - you can add `resolutions` with acorn 8.x to your package.json.

**Comment by marcelmusiol** [comment]
Did anyone find a working solution for Nextjs? I tried adding adding `resolutions` with acorn 8.x to package.json and using the `@babel/plugin-proposal-optional-chaining` plugin, but no success so far.

**Comment by talentlessguy** [comment]
Should work with:

```json
"@react-three/fiber": "^6.0.1",
"@react-three/postprocessing": "^2.0.1",
"next": "^10.1.2",
"postprocessing": "^6.21.2",
"react": "^17.0.2",
"react-dom": "^17.0.2",
"three": "^0.127.0"
```

@kyleawayan @nathanbabcock @vinkovsky @sargunv @MichalPrzybys-TomTom @marcelmusiol lmk if the issue is still present

**Comment by denkravchu** [comment]
still the same error with nuxt js

**Comment by drcmda** [comment]
> still the same error with nuxt js

is it latest? can you go back a few versions, at least before ssr was merged? maybe it's that.

**Comment by williaster** [comment]
I still get this error with latest everything:

```json
"@react-three/drei": "^9.17.0",
"@react-three/fiber": "^8.0.27",
"@react-three/postprocessing": "^2.4.6",
"next": "12.2.1",
"postprocessing": "^6.28.1",
"react": "18.2.0",
"react-dom": "18.2.0",
"three": "^0.142.0"
```

also tried adding `next-transpile-modules` to transpile the build `node_modules` but no dice.

```
error - /Users/christopher-williams/dev/r3f-spaceship/node_modules/postprocessing/build/postprocessing.js:1405
      this.fragmentShader = effect_default.replace(EffectShaderSection.FRAGMENT_HEAD, shaderParts.get(EffectShaderSection.FRAGMENT_HEAD) ?? "").replace(EffectShaderSection.FRAGMENT_MAIN_UV, shaderParts.get(EffectShaderSection.FRAGMENT_MAIN_UV) ?? "").replace(EffectShaderSection.FRAGMENT_MAIN_IMAGE, shaderParts.get(EffectShaderSection.FRAGMENT_MAIN_IMAGE) ?? "");
                                                                                                                                          ^

SyntaxError: Unexpected token '?'
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/Users/christopher-williams/dev/r3f-spaceship/node_modules/@react-three/postprocessing/dist/index.cjs:5:22)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10) {
  page: '/'
}
```

**Comment by K-JAX** [comment]
In the off chance this helps anybody who finds their way here - I got this error when I tried to use the code from the r3f docs page for [Depth of Field](https://docs.pmnd.rs/react-postprocessing/effects/depth-of-field) not understanding that in order to use the DepthOfField component, it must be placed in the EffectComposer component as stated in the react-postprocessing README.
TLDR In other words, my fix was:
```
<Canvas>
    <EffectComposer>
        <DepthOfField />
        {/* The rest of your scene */}
    </EffectComposer>
</Canvas>
```

**Comment by ZowieTao** [comment]
In the Next.js 13 project, global server-side rendering (SSG) is enabled by default, and the <DepthOfField> component is for client-side rendering (CSR) effects and cannot be processed during server-side rendering. This can cause build errors. To fix this, you can move the code that uses the <DepthOfField> component into a client-rendered environment to ensure that it is only executed on the client side.
```javascript
const DepthOfField = dynamic(
  () =>
    import("@react-three/postprocessing").then((module) => module.DepthOfField),
  {
    ssr: false,
  },
)
const EffectComposer = dynamic(
  () =>
    import("@react-three/postprocessing").then(
      (module) => module.EffectComposer,
    ),
  {
    ssr: false,
  },
)
```

**Comment by Uilocalnotification** [comment]
> On the off chance this helps anybody who finds their way here - I got this error when I tried to use the code from the r3f docs page for [Depth of Field](https://docs.pmnd.rs/react-postprocessing/effects/depth-of-field) not understanding that in order to use the DepthOfField component, it must be placed in the EffectComposer component as stated in the react-postprocessing README. TLDR In other words, my fix was:
> 
> ```
> <Canvas>
>     {/* The rest of your scene */}
>     <EffectComposer>
>         <DepthOfField />
>     </EffectComposer>
> </Canvas>
> ```





https://stackdev.space/questions/the-issue-with-srgb-encoding-persisting-in-threeeffectcomposer-has-not-been-resolved
